src/Microsoft.MobileBlazorBindings.Authentication/Http/AuthorizationMessageHandler.cs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.MobileBlazorBindings.Authentication
+{
+    /// <summary>
+    /// A <see cref="DelegatingHandler"/> that attaches access tokens to outgoing <see cref="HttpResponseMessage"/> instances.
+    /// Access tokens will only be added when the request URI is within one of the base addresses configured using
+    /// <see cref="ConfigureHandler(IEnumerable{string}, IEnumerable{string})"/>.
+    /// </summary>
+    /// <remarks>
+    /// Since message handlers have their own scope to resolve from, <see cref="RegisterTokenProvider(IEnumerable{string}, IAccessTokenProvider)"/>
+    /// needs to be called to register the correct <see cref="IAccessTokenProvider"/> for the provided base URL.
+    /// </remarks>
+    public class AuthorizationMessageHandler : DelegatingHandler
+    {
+        private static readonly ConcurrentDictionary<string, IAccessTokenProvider> TokenProviders = new ConcurrentDictionary<string, IAccessTokenProvider>();
+
+        private Uri[] _authorizedUris;
+        private AccessTokenRequestOptions _tokenOptions;
+
+        /// <inheritdoc />
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (request == null)
+                throw new ArgumentNullException(nameof(request));
+
+            var now = DateTimeOffset.Now;
+            if (_authorizedUris == null)
+            {
+                throw new InvalidOperationException($"The '{nameof(AuthorizationMessageHandler)}' is not configured. " +
+                    $"Call '{nameof(AuthorizationMessageHandler.ConfigureHandler)}' and provide a list of endpoint urls to attach the token to.");
+            }
+
+            var authorizedUri = _authorizedUris.SingleOrDefault(uri => uri.IsBaseOf(request.RequestUri));
+
+            if (authorizedUri != null)
+            {
+                // need to intern it to make sure that reference equals succeeds.
+                var key = string.Intern(authorizedUri.ToString());
```

Comment:
Why do you care about reference equality?

Reply:
You are right, I was originally using ConditionalWeakTable, that uses reference equality. Removed.